### DISCUSSION TITLE
Problem mit der Kommunikation JK-PB2A16S15P

### DISCUSSION THREAD

**Discussion by stamartin** [hoylabs/OpenDTU-OnBattery#1258] [Q&A] [answered]
Hallo,
erstmal vielen Dank an alle, die diese Projekt entwickeln.

Ich habe mein BKW mit 2kWp, BlueSolar 150/45, HMS1600 und US2000C zum Jahreswechsel aufgebaut. Inzwischen habe ich aber festgestellt, dass 2,4kWh eindeutig zu klein sind. Erstens reicht die Kapazität nicht und zweitens sind die Störme zu groß (obwohl ich so weit es geht dagegen steuere). Deswegen habe ich mir 16 289er Zellen und das EEL mit JK-BMS gekauft und zusammengebaut.
Jetzt zu meinem Problem: Ich bekomme nicht die Kommunikation zwischen dem ESP32 und dem JK-BMS mittels RS485 hin. Ich habe streng nach Wiki ein HW-0519 (optisch das gleiche wie das "DEBO TTL-RS485" von Reichelt) mit 4 Adern (3,3V, GND, GPIO 41,42) verbunden und auf der anderen Seite mit einem RJ45-Adapter (1 B, 2 A, 3 GND, ausprobiert habe ich auch 8,7,6) verbunden. Das Netzwerkkabel in die linke Buchse gesteckt. Das Protokoll ist 0 (4G-GPS...).
Im ESP habe ich die GPIOs folgendermaßen belegt:
Battery | rx | 41 
          rxen | 40 
              tx | 42 
          txen | 39
In den Batterie-Einstellungen auf JK und TTl-UART an der MCU.
In den Weiten des Internets habe ich etwas von verdrehter Pin-Belegung gelesen, deswegen habe ich TX und RX in jeder Konstellation ausprobiert.
Auf der Platine blinkt die Tx-LED im eingestellten Takt. Mehr passiert leider nicht...
Als RS485-Adresse habe ich 0 und 1 ausprobiert.
Aus Verzweifelung habe ich das Can-Protokoll auf Pylontech eingestellt und damit kann ich die wichtigsten Daten abgreifen.
Die Buchse und das BMS würde ich ausschließen, da ich zwischen den A & B Pins 140 Ohm messe und eine Verbindung mit dem PC bekomme. Das Kabel habe ich auch testweise ausgetauscht.

Hat diese Kombination jemand im Einsatz?
Hat jemand eine Idee?

**Comment by Manos1966** [comment]
Ich habe ein Mal die Kommunikation mit JK BMS problemlos gemacht ueber RS485-TTL (UART).
Mehr Infos unter https://opendtu-onbattery.net/hardware/jkbms/models_b/#protocols
( RS485-TTL (UART) Jack Pin-Out)

**Comment by schlimmchen** [comment]
Außer, dass du nicht verstanden hast, dass man rxen und txen an diesem RS485-Adapter nicht hat und daher auch nicht belegen musst im pin mapping, insb. wenn du "TTL-UART an MCU" auswählst, fällt mir nichts grob falsches auf. Ich habe so ein Inverter-BMS schon verbunden und auf dessen Grundlage die [Doku](https://opendtu-onbattery.net/hardware/jkbms/models_pb/) geschrieben. Allerdings habe ich das OpenDTU Fusion Board genommen und an den integrierten RS485 Transceiver angeschlossen. Das sollte aber keinen entscheidenden Unterschied machen.

Hast du wirklich alles an Kabeln zwischen dem BMS und dem Board mit den Buchsen angeschlossen?

GPIO41 und 42 sollten auch okay sein, egal ob ESP32 oder ESP32-S3. Und du hast die Pins nicht doppelt vergeben? Und du hast das Device-Profile auch tatsächlich nicht nur hochgeladen, sondern auch ausgewählt?

Was steht denn in der Konsole, wenn du verbose logging einschaltest?

Hast du die Drähte verlötet oder gesteckt?

**Reply by stamartin** [reply]
> Hast du wirklich alles an Kabeln zwischen dem BMS und dem Board mit den Buchsen angeschlossen?

Die Batterie habe ich nach der Anleitung von EEL aufgebaut. Da waren alle Kabel dabei. Ich konnte auch zwischen 1 und 2 (bzw 7 und8) 140Ohm auf der Rückseite der Buchse messen. Das Kabel konnte ich bis zum BMS durchklingeln.
Es ist ein ESP32-S3.

"name": "OpenDTU v3",
        "cmt": {
            "clk": 4,
            "cs": 7,
            "fcs": 6,
            "sdio": 5,
            "gpio2": -1,
            "gpio3": -1
        },
        "led": {
            "led0": 17,
            "led1": 18
        },
        "victron": {
            "rx": 15,
            "tx": 16
        },
        "battery": {
            "rx": 41,
            "tx": 42,
            "rxen": 40,
            "txen": 39
        }

Das Device-Profil läuft wunderbar mit der Pylontech. Und das JK läuft auch mit CAN auf den beiden Pins. Alle Verbindungen sind mit neuen Jumperkabel verbunden. Der ESP sitzt auf ein BreakoutBoard [https://www.amazon.de/ESP32-Breakout-WROOM1-DevKitC-Development/dp/B0D1K8RMXB?source=ps-sl-shoppingads-lpcontext&ref_=fplfs&psc=1&smid=A2OYQMAUQB5NOO]

21:10:05.070 > [ 257849.445] JK BMS: request for data sent
21:10:05.119 > [ 257849.447] JK BMS: busy waiting for or reading a message from the BMS
21:10:05.442 > [ 257849.456] JK BMS: waiting for poll interval to elapse
21:10:05.678 > Middle missing
21:10:05.741 > Request retransmit: 2
21:10:05.804 > Middle missing
21:10:05.885 > Request retransmit: 2

> Has a 120 ohm termination resistor with the option to enable it by shorting R0

Das habe ich gerade in den Beschreibungen gefunden. Ich werde morgen probieren den R0 kurzzuschließen. Evtl. bringt das den Erfolg.

**Reply by stamartin** [reply, marked as answer]
Ich habe R0 kurzgeschlossen. Danach gab es einmal eine Verbindung. Beim wackeln an dem RJ45/Stiftadapter wieder eine Verbindung. Also den Adapter getauscht und siehe da. Alles läuft.
Vielen Dank für die Hilfe!
